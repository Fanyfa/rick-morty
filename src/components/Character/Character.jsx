import { string } from 'prop-types';
import { withAppContext } from 'providers/App';
import React, { Component } from 'react';
import styles from './styles.module.scss';

class Character extends Component {
  static propTypes = {
    name: string.isRequired,
    status: string.isRequired,
    species: string.isRequired,
    gender: string.isRequired,
    image: string.isRequired,
    origin: string,
    type: string,
    location: string,
    episodes: string,
  };

  static defaultProps = {
    origin: undefined,
    lastLocation: undefined,
    episodes: undefined,
    type: undefined,
  };

  render() {
    const { name, image, status, species, gender, origin, type, location, episodes } = this.props;
    return (
      <div className={styles.card}>
        <div style={{ backgroundImage: `url('${image}')` }} alt={name} className={styles.image} />
        <div className={styles.caption}>
          <div className={styles.name}>{name}</div>
          <div className={styles.species}>{species}</div>
          <ul className={styles.details}>
            <li>
              <span className={styles.detailsItem}>Status:</span>
              {status}
            </li>
            <li>
              <span className={styles.detailsItem}>Gender:</span>
              {gender}
            </li>
            {origin ? (
              <li>
                <span className={styles.detailsItem}>Origin:</span>
                {origin}
              </li>
            ) : (
              ''
            )}
            {type ? (
              <li>
                <span className={styles.detailsItem}>Type:</span>
                {type}
              </li>
            ) : (
              ''
            )}
            {location ? (
              <li>
                <span className={styles.detailsItem}>Location:</span>
                {location}
              </li>
            ) : (
              ''
            )}
            {episodes ? (
              <li>
                <span className={styles.detailsItem}>Appears in episodes:</span>
                {episodes}
              </li>
            ) : (
              ''
            )}
          </ul>
        </div>
      </div>
    );
  }
}

export default withAppContext(Character);
