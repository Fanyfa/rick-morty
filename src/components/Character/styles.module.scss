.card {
  display: inline-block;
  width: 300px;
  padding: 1em;
  border-radius: 15px;
  margin: 10px;
  background: linear-gradient(20deg, #191919 0%, #100b32 33%, #5c0249 100%);
  text-align: left;
  box-shadow: 0px 5px 20px -10px #111111;
  position: relative;
  transition: 0.4s;

  &:hover {
    transform: translateY(-10px) rotate(-3deg);
  }

  .image {
    background-position: top center;
    background-size: cover;
    background-repeat: no-repeat;
    width: 100%;
    height: 280px;
  }

  .caption {
    background-color: rgba(255, 255, 255, 0.65);
    padding: 1em;
    position: relative;
    border-radius: 0 0 3px 3px;

    .name {
      text-align: center;
      font-size: 20px;
      font-weight: bold;
    }

    .species {
      position: absolute;
      top: 0;
      right: 1em;
      -webkit-transform: translateY(-50%);
      transform: translateY(-50%);
      color: #ffffff;
      text-transform: uppercase;
      font-family: 'Open Sans Condensed', 'Open Sans', helvetica, sans-serif;
      letter-spacing: 0.1em;
      padding: 0.25em;
      line-height: 1;
      border-radius: 2px;
      background: #5c0249;
    }

    .details {
      margin: 10px;

      .detailsItem {
        margin-right: 10px;
        font-weight: bold;
      }
    }
  }
}
