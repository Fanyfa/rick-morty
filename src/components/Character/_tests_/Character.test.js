import Character from 'components/Character/';
import React from 'react';
import renderer from 'react-test-renderer';

const NAME = 'Beth';
const STATUS = 'Alive';
const SPECIES = 'Human';
const GENDER = 'Female';
const IMAGE = 'https://rickandmortyapi.com/api/character/avatar/3.jpeg';
const ORIGIN = 'Earth';
const TYPE = 'Human';
const LOCATION = 'Earth';
const EPISODES = '1, 2, 3';

describe('<Character>', () => {
  it('renders', () => {
    const tree = renderer
      .create(<Character name={NAME} status={STATUS} species={SPECIES} gender={GENDER} image={IMAGE} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {origin}', () => {
    const tree = renderer
      .create(
        <Character
          name={NAME}
          status={STATUS}
          species={SPECIES}
          gender={GENDER}
          image={IMAGE}
          origin={ORIGIN}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {type}', () => {
    const tree = renderer
      .create(
        <Character name={NAME} status={STATUS} species={SPECIES} gender={GENDER} image={IMAGE} type={TYPE} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {location}', () => {
    const tree = renderer
      .create(
        <Character
          name={NAME}
          status={STATUS}
          species={SPECIES}
          gender={GENDER}
          image={IMAGE}
          location={LOCATION}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {episodes}', () => {
    const tree = renderer
      .create(
        <Character
          name={NAME}
          status={STATUS}
          species={SPECIES}
          gender={GENDER}
          image={IMAGE}
          episodes={EPISODES}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
