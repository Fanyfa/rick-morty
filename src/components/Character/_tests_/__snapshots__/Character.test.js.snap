// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Character> renders 1`] = `
<div
  className="card"
>
  <div
    alt="Beth"
    className="image"
    style={
      Object {
        "backgroundImage": "url('https://rickandmortyapi.com/api/character/avatar/3.jpeg')",
      }
    }
  />
  <div
    className="caption"
  >
    <div
      className="name"
    >
      Beth
    </div>
    <div
      className="species"
    >
      Human
    </div>
    <ul
      className="details"
    >
      <li>
        <span
          className="detailsItem"
        >
          Status:
        </span>
        Alive
      </li>
      <li>
        <span
          className="detailsItem"
        >
          Gender:
        </span>
        Female
      </li>
      
      
      
      
    </ul>
  </div>
</div>
`;

exports[`<Character> when {episodes} 1`] = `
<div
  className="card"
>
  <div
    alt="Beth"
    className="image"
    style={
      Object {
        "backgroundImage": "url('https://rickandmortyapi.com/api/character/avatar/3.jpeg')",
      }
    }
  />
  <div
    className="caption"
  >
    <div
      className="name"
    >
      Beth
    </div>
    <div
      className="species"
    >
      Human
    </div>
    <ul
      className="details"
    >
      <li>
        <span
          className="detailsItem"
        >
          Status:
        </span>
        Alive
      </li>
      <li>
        <span
          className="detailsItem"
        >
          Gender:
        </span>
        Female
      </li>
      
      
      
      <li>
        <span
          className="detailsItem"
        >
          Appears in episodes:
        </span>
        1, 2, 3
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Character> when {location} 1`] = `
<div
  className="card"
>
  <div
    alt="Beth"
    className="image"
    style={
      Object {
        "backgroundImage": "url('https://rickandmortyapi.com/api/character/avatar/3.jpeg')",
      }
    }
  />
  <div
    className="caption"
  >
    <div
      className="name"
    >
      Beth
    </div>
    <div
      className="species"
    >
      Human
    </div>
    <ul
      className="details"
    >
      <li>
        <span
          className="detailsItem"
        >
          Status:
        </span>
        Alive
      </li>
      <li>
        <span
          className="detailsItem"
        >
          Gender:
        </span>
        Female
      </li>
      
      
      <li>
        <span
          className="detailsItem"
        >
          Location:
        </span>
        Earth
      </li>
      
    </ul>
  </div>
</div>
`;

exports[`<Character> when {origin} 1`] = `
<div
  className="card"
>
  <div
    alt="Beth"
    className="image"
    style={
      Object {
        "backgroundImage": "url('https://rickandmortyapi.com/api/character/avatar/3.jpeg')",
      }
    }
  />
  <div
    className="caption"
  >
    <div
      className="name"
    >
      Beth
    </div>
    <div
      className="species"
    >
      Human
    </div>
    <ul
      className="details"
    >
      <li>
        <span
          className="detailsItem"
        >
          Status:
        </span>
        Alive
      </li>
      <li>
        <span
          className="detailsItem"
        >
          Gender:
        </span>
        Female
      </li>
      <li>
        <span
          className="detailsItem"
        >
          Origin:
        </span>
        Earth
      </li>
      
      
      
    </ul>
  </div>
</div>
`;

exports[`<Character> when {type} 1`] = `
<div
  className="card"
>
  <div
    alt="Beth"
    className="image"
    style={
      Object {
        "backgroundImage": "url('https://rickandmortyapi.com/api/character/avatar/3.jpeg')",
      }
    }
  />
  <div
    className="caption"
  >
    <div
      className="name"
    >
      Beth
    </div>
    <div
      className="species"
    >
      Human
    </div>
    <ul
      className="details"
    >
      <li>
        <span
          className="detailsItem"
        >
          Status:
        </span>
        Alive
      </li>
      <li>
        <span
          className="detailsItem"
        >
          Gender:
        </span>
        Female
      </li>
      
      <li>
        <span
          className="detailsItem"
        >
          Type:
        </span>
        Human
      </li>
      
      
    </ul>
  </div>
</div>
`;
